source/client/process_impl.cc

```diff
@@ -227,8 +227,15 @@ class NighthawkServerInstance : public Envoy::Server::Instance {
   envoy::config::bootstrap::v3::Bootstrap& bootstrap() override {
     PANIC("NighthawkServerInstance::bootstrap not implemented");
   }
+  void setServerFactoryContext(
```

Comment:
Having a `setFoo` method implies that one chan call it multiple times (i.e. swap the Foo instance during its lifetime).

I am assuming that we don't have such use case and also that such usecase would not be suppoted (thread safety, etc.).

Can we keep the API minimal by setting the factory context at construction time?

Reply:
I don't see a way to achieve this at construction time with the current design, because the two classes both need to have references to each other.
-  The `NighthawkServerInstance` is passed to the constructor of the `NighthawkServerFactoryContext`, implying `NighthawkServerInstance` needs to be created first. 
- Now adding `setServerFactoryContext` to come back around and provide the `NighthawkServerInstance` to the `NighthawkServerFactoryContext` since we didn't have it at construction time.

What if we also have an assertion that only allows setFactoryContext to be called once? And maybe rename it `initializeFactoryContext`?

Another possibility might be an even bigger super-object that implements both `ServerFactoryContext` and `ServerInstance`.

Reply:
This just in: `Envoy::Server::Configuration::ServerFactoryContext` methods are exactly a subset of `Envoy::Server::Configuration::Instance` methods. I will see what I can do -- the super-object we seek may just be `Envoy::Server::Configuration::Instance` itself.

Reply:
Good news, thank you for looking deeper.

Reply:
I found a way to eliminate the circular dependency.

The job of `NighthawkServerInstance` and `NighthawkServerFactoryContext` is to return references to various Envoy objects in Nighthawk like the `Api` object.

Because of recent Envoy changes, `NighthawkServerInstance` needed to start implementing `serverFactoryContext()` instead of the `PANIC` placeholder. The easiest way to implement `serverFactoryContext()` is to store a reference to a prebuilt `ServerFactoryContext`.

Unfortunately in the existing code,  `NighthawkServerFactoryContext` was already storing a reference to `NighthawkServerInstance` and delegating a lot of method calls to it. So there would be mutual references between and  `NighthawkServerFactoryContext` and `NighthawkServerInstance`. I don't know a way to initialize two classes that mutually reference each other, without opening the door to setters and pointers.

Fortunately the delegation in `NighthawkServerFactoryContext` was unnecessary. I updated `NighthawkServerFactoryContext` to store its own references to the objects it had been getting from inside the `NighthawkServerInstance`, and eliminated its reference to the `NighthawkServerInstance`.